pkg/roachprod/cloud/cluster_cloud.go

```diff
@@ -311,31 +311,50 @@ func ListCloud(l *logger.Logger, options vm.ListOptions) (*Cloud, error) {
 	return cloud, providerErr
 }
 
+type ClusterCreateOpts struct {
+	// Nodes indicates how many nodes in the cluster should be created with the
+	// respective CreateOpts and ProviderOpts.
+	Nodes                 int
+	CreateOpts            vm.CreateOpts
+	ProviderOptsContainer vm.ProviderOptionsContainer
+}
+
 // CreateCluster TODO(peter): document
-func CreateCluster(
-	l *logger.Logger,
-	nodes int,
-	opts vm.CreateOpts,
-	providerOptsContainer vm.ProviderOptionsContainer,
-) error {
-	providerCount := len(opts.VMProviders)
-	if providerCount == 0 {
-		return errors.New("no VMProviders configured")
+// opts is a slice of all node VM specs to be provisioned for the cluster. Generally,
+// non uniform VM specs are not supported for a CRDB cluster, but we often want to provision
+// an additional "workload node". This node often times does not need the same CPU count as
+// the rest of the cluster. i.e. it is overkill for a 3 node 32 CPU cluster to have a 32 CPU
+// workload node, but a 50 node 8 CPU cluster might find a 8 CPU workload node inadequate.
+func CreateCluster(l *logger.Logger, opts []ClusterCreateOpts) error {
+	// Keep track of the total number of nodes created, as we append all cluster names
+	// with the node count.
+	var nodesCreated int
+	vmName := func(name string) string {
+		nodesCreated++
+		return vm.Name(name, nodesCreated)
 	}
+	for _, o := range opts {
```

Comment:
Nit: this would result in an unnecessary copy of `ClusterCreateOpts`. These structs aren't deeply nested nor is this code on a critical path. Nevertheless, it's something to keep in mind especially if the number of created nodes grows by orders of magnitude. There is an extra copy which happens when passing formals into actuals at a call-site, so you might as well pass `[]*ClusterCreateOpts`.

Reply:
Ahh, fixed!